torch/csrc/jit/runtime/profiling_record.h

```diff
@@ -8,12 +8,127 @@
 #include <torch/csrc/jit/ir/ir.h>
 
 #include <list>
+#include <map>
 #include <vector>
 
+// We would like to assign each position/axis of a tensor an abstract size
+// * For each `tensor` we have a profiled `Value` of a `TensorType` describing the properties of the `tensor`.
+// * `TensorType` has a property called `symbolic_sizes_` to describe observed `tensor.sizes()`
+// * `symbolic_sizes_` is a vector of abstract sizes (or `std::vector<ShapeSymbol>`) where
+//   * `ShapeSymbol`at `symbolic_sizes_[i]`  describes the size value (`Dimension`) at `tensor.sizes()[i]`
+// * We may see the same `Dimension` at different positions `i` in `tensor.sizes()` or even in different `tensor`
+//   * First, we would like associate the same `ShapeSymbol` to the same `Dimension` across **one** profiling execution or run 	of a TorchScript function.
+//     * The same `ShapeSymbol`s in different positions of `symbolic_shapes_` in possibly different `TensorType`s (i.e. `TensorType`s for different profiled values) form an implicit set. The elements of such a set are called *dimension locations*. 
+//     * These sets allow us to track how the shapes of input arguments of some operation relate to operation's output shapes as the input and output shapes might share the same `ShapeSymbol`s
+// * For **every** profiling run, we would like to maintain the invariant that *the same `ShapeSymbol` is always associated with the same `Dimension`*.
+// * To maintain this invariant we merge the profiling information from all profiling runs, 
+//   * For every two runs, we iterate over all `symbic_shapes_`  and compare their `ShapeSymbol`s in the same position. 
+//     * if we observe that for every dimension location that has the`ShapeSymbol S1`  in run #1 there is **only one** `ShapeSymbol S2` in the same dimension location in run #2, we conclude that the invariant holds.
+//     * However, if we observe some dimension locations in run #2 have `ShapeSymbol S2` and the other ones have `ShapeSymbol S3` we would like to partition the virtual set of dimension locations associated with `ShapeSymbol S1` into two new subsets, so the invariant holds.
+//     * The partitioning works by assigning a new symbol to the dimension locations (associated with `ShapeSymbol S1`) that have `ShapeSymbol S2` and another new symbol to the dimension locations that have `ShapeSymbol S3`. In other words,
```

Comment:
Thanks for adding this comment, it makes it much clearer what is happening in the merging algorithm.